packages/node/src/login/oidc/redirectHandler/AuthCodeRedirectHandler.ts

```diff
@@ -26,16 +26,141 @@
 
 import { inject, injectable } from "tsyringe";
 import {
+  IClient,
   IClientRegistrar,
+  IIssuerConfig,
   IIssuerConfigFetcher,
   IRedirectHandler,
   ISessionInfo,
   ISessionInfoManager,
   IStorageUtility,
-  NotImplementedError,
 } from "@inrupt/solid-client-authn-core";
 import { URL } from "url";
-import { fetchType } from "../../../authenticatedFetch/fetchFactory";
+import { IdTokenClaims, Issuer, TokenSet } from "openid-client";
+import { JWK } from "jose";
+import { configToIssuerMetadata } from "../IssuerConfigFetcher";
+import {
+  buildBearerFetch,
+  buildDpopFetch,
+  fetchType,
+} from "../../../authenticatedFetch/fetchFactory";
+
+type OidcContext = {
+  sessionId: string;
+  issuerConfig: IIssuerConfig;
+  codeVerifier: string;
+  redirectUri: string;
+  dpop: boolean;
+};
+
+/**
+ * Based on the provided state, this looks up contextual information stored
+ * before redirecting the user to the OIDC issuer.
+ * @param oauthState The state (~ correlation ID) of the OIDC request
+ * @param storageUtility
+ * @param configFetcher
+ * @returns Information stored about the client issuing the request
+ */
+async function retrieveContextFromStorage(
+  oauthState: string,
+  storageUtility: IStorageUtility,
+  configFetcher: IIssuerConfigFetcher
+): Promise<OidcContext> {
+  try {
+    // Since we throw if not found, the type assertion are ok too
+    const storedSessionId = (await storageUtility.getForUser(
+      oauthState,
+      "sessionId",
+      {
+        errorIfNull: true,
+      }
+    )) as string;
+
+    const [
+      issueIri,
+      codeVerifier,
+      storedRedirectIri,
+      dpop,
+    ] = (await Promise.all([
+      storageUtility.getForUser(storedSessionId, "issuer", {
+        errorIfNull: true,
+      }),
+      storageUtility.getForUser(storedSessionId, "codeVerifier", {
+        errorIfNull: true,
+      }),
+      storageUtility.getForUser(storedSessionId, "redirectUri", {
+        errorIfNull: true,
+      }),
+      storageUtility.getForUser(storedSessionId, "dpop", { errorIfNull: true }),
+    ])) as string[];
+
+    // Unlike openid-client, this looks up the configuration from storage
+    const issuerConfig = await configFetcher.fetchConfig(issueIri);
+    return {
+      sessionId: storedSessionId,
+      codeVerifier,
+      redirectUri: storedRedirectIri,
+      issuerConfig,
+      dpop: dpop === "true",
+    };
+  } catch (e) {
+    throw new Error(
+      `Could not retrieve the OIDC context from storage for request associated to state ${oauthState}: ${e.toString()}`
+    );
+  }
+}
+
+async function saveSessionInfoInStorage(
```

Comment:
We don't save session info anywhere else apart from Storage (right?), so can this name be simplified, i.e.:
```suggestion
async function saveSessionInfo(
```

Reply:
687516df762ab9f8dbbaae9605bf250ac66de6dd